### PR TITLE
refactor(all): use nvim-treesitter and just completion source

### PR DESCRIPTION
**Pull request by steelsojka**
Note, this removes all features that are not related to the completion source. These features will be ported over to a module in nvim-treesitter in the near future.

### CODE REVIEW

File: lua/completion-treesitter/source.lua [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  end
+
+  return current or nil
+end
+
+local function prepare_match(match, kind)
+  local matches = {}
+
+  if match.node then
+    table.insert(matches, { kind = M.match_kinds[kind] or '', full = match.node, def = match.node })
```
Comment (vigoux) [inline lua/completion-treesitter/source.lua]: I am not sure we need both `full` and `def` as they now are the same

File: lua/completion-treesitter/source.lua [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        local node = match.def
+        local node_scope = get_smallest_context(node)
+        local start_line_node, _, _= node:start()
+
+        local node_text = ts_utils.get_node_text(node, bufnr)[1]
+
+        -- Only consider items in current scope, and not already met
+        local score = score_func(prefix, node_text)
+
+        if score < #prefix / 2 and node_text
```
Comment (vigoux) [inline lua/completion-treesitter/source.lua]: Is the filtering still handled by us ? I have not followed completion-nvim dev recently but did it changed ?

**Review by vigoux — COMMENTED** [review]
Thank you sooo much for this !
Really nice contribution and really nice work !

File: lua/completion-treesitter/source.lua (hunk 1 above)
Comment (steelsojka) [reply lua/completion-treesitter/source.lua]: I thought the same thing. I was going to remove it, but forgot.

File: lua/completion-treesitter/source.lua (hunk 2 above)
Comment (steelsojka) [reply lua/completion-treesitter/source.lua]: I'll double check.

**Comment by steelsojka** [comment]
@vigoux The one feature I couldn't get was adding the `full_text` of the declaration. It required a custom query. Since this uses the local queries, it wasn't possible to get it. Any ideas?

**Comment by vigoux** [comment]
Hmm.... One way I see is modifying locals to include that to... Another way is to get the content of the line where that node is defined and use it as full text...

**Comment by steelsojka** [comment]
I'll just get the full line where the node is used. That seems like the easiest way.

**Comment by vigoux** [comment]
And actually that is kind of the closest from default omnifunc completion

**Comment by vigoux** [comment]
Now that this is out of the way, we should try and define how the locals should behave to give us the most out of the syntax tree, I was planning to add some kind of type based completion, but I don't know if it is possible though with the current implementation

**Comment by steelsojka** [comment]
Do you have an example of type based completion?

**Comment by vigoux** [comment]
Like completing the fields of a C struct, or completing with the keys of a table in Lua when possible, or enum variants

**Comment by steelsojka** [comment]
Ah I see. And this would only be possible if the type was defined within the same file/tree? Is there anything this PR should do to allow for this in the future?

**Comment by vigoux** [comment]
Yes, only for types defined in the current file.
As it is quite speculative, I think that we should first include that massive change, then think about it. 

I will tag the current version as 0.1 and that new one, after you PR is merged as 1.0 as it is a breaking change. We will have to notify in the README that you PR is breaking some thing but they will be available in `nvim-treesitter`.

**Comment by steelsojka** [comment]
Ok. Sounds good. I'll update the README.

**Comment by vigoux** [comment]
Sorry for the delay, I would like to try this PR before merging, but I really don't have that much time now, I will soon hopefully...

**Comment by vigoux** [comment]
Really nice addition, thank you for that (again) !

**Pull request merged**